### PR TITLE
Fix support for Knative services

### PR DESCRIPTION
**Pull request by mark-burnett**
This re-instates support for image substitution with Knative services, which appears to have been broken when the fix for not substituting images in CRDs was implemented in #3833

### CODE REVIEW

**Comment by mark-burnett** [comment]
Hey @briandealwis, thanks for taking a look. I think this lookup is effectively also O(k), since the cost is bounded by the length of the whitelist rather than O(n) on the input.

That said, after benchmarking both, the map is faster even when only looking for the first element in the list.  My only guess is that this is related to string comparison being possibly slower than hashing the object.

I'll follow up with an update.

```
Benching Map
  avg 106ns;  min 54ns;  p50 101ns;  max 582µs;
  p90 105ns;  p99 136ns;  p999 399ns;  p9999 13.4µs;
       54ns [ 4123649] ██████████
      100ns [15717538] ████████████████████████████████████████
      150ns [   36341] 
      200ns [   26603] 
      250ns [   19593] 
      300ns [      58] 
      350ns [   56929] 
      400ns [   12095] 
      450ns [    2020] 
      500ns+[    5174] 

Benching List
  avg 126ns;  min 76ns;  p50 122ns;  max 248µs;
  p90 124ns;  p99 140ns;  p999 207ns;  p9999 15.7µs;
       76ns [       1] 
       80ns [       0] 
      100ns [ 1734240] ███▌
      120ns [18029092] ████████████████████████████████████████
      140ns [  151172] 
      160ns [   22589] 
      180ns [   35973] 
      200ns [   11689] 
      220ns [    3155] 
      240ns+[   12089] 

Benching Indexed List
  avg 138ns;  min 85ns;  p50 131ns;  max 85.2µs;
  p90 142ns;  p99 186ns;  p999 262ns;  p9999 18.1µs;
       85ns [       1] 
      100ns [       0] 
      120ns [17496609] ████████████████████████████████████████
      140ns [ 2094673] ████▌
      160ns [  192674] 
      180ns [   78874] 
      200ns [   68662] 
      220ns [   37147] 
      240ns [   10704] 
      260ns+[   20656] 

Benching Indexed List Valid Only
  avg 121ns;  min 62ns;  p50 111ns;  max 262µs;
  p90 132ns;  p99 146ns;  p999 262ns;  p9999 18.3µs;
       62ns [       1] 
       80ns [       0] 
      100ns [16863182] ████████████████████████████████████████
      120ns [ 1926083] ████▌
      140ns [ 1067329] ██▌
      160ns [   67193] 
      180ns [   25375] 
      200ns [   18533] 
      220ns [    7406] 
      240ns+[   24898] 

```

File: pkg/skaffold/deploy/kubectl/visitor.go [hunk 1]
```diff
@@ -20,17 +20,42 @@ import (
...
+		Kind:  "CronJob",
+	}: true,
+	{
+		Group: "batch",
+		Kind:  "Job",
+	}: true,
+	{
+		Group: "serving.knative.dev",
+		Kind:  "Service",
+	}: true,
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/visitor.go:58]: This is a bit difficult to read.  It should also have a doc string.  How about:
```suggestion
// transformableWhitelist is the set of kinds that can be transformed by Skaffold.
var transformableWhitelist = map[apimachinery.GroupKind]bool{
	{Group: "", Kind: "Pod"}:                        true,
	{Group: "apps", Kind: "DaemonSet"}:              true,
	{Group: "apps", Kind: "Deployment"}:             true,
	{Group: "apps", Kind: "ReplicaSet"}:             true,
	{Group: "apps", Kind: "StatefulSet"}:            true,
	{Group: "batch", Kind: "CronJob"}:               true,
	{Group: "batch", Kind: "Job"}:                   true,
	{Group: "serving.knative.dev", Kind: "Service"}: true,
}
```

Comment (mark-burnett) [reply pkg/skaffold/deploy/kubectl/visitor.go]: Thanks, applied!

File: pkg/skaffold/deploy/kubectl/visitor.go [hunk 2]
```diff
@@ -20,17 +20,20 @@ import (
...
+var transformableWhitelist = map[apimachinery.GroupKind]bool{
+	{Group: "", Kind: "Pod"}:                        true,
+	{Group: "apps", Kind: "DaemonSet"}:              true,
+	{Group: "apps", Kind: "Deployment"}:             true,
+	{Group: "apps", Kind: "ReplicaSet"}:             true,
+	{Group: "apps", Kind: "StatefulSet"}:            true,
+	{Group: "batch", Kind: "CronJob"}:               true,
+	{Group: "batch", Kind: "Job"}:                   true,
+	{Group: "serving.knative.dev", Kind: "Service"}: true,
+}
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/visitor.go:36]: Oops, sorry — there's a stray `}`

Comment (mark-burnett) [reply pkg/skaffold/deploy/kubectl/visitor.go:36]: Ah, I should have noticed, sorry :)

**Pull request merged**